commonknowledge/join
# feat: wrap join request handling in an exclusive lock to prevent race conditions

## Description
Despite vain hopes that race conditions wouldn't happen, they are happening.
This is an attempt to prevent them with a solid, theoretical approach: exclusive locking.
When a join request is received, use PHP's `flock` – file based locking – to make sure
that only one thread is processing this join flow session.

The lock is specific to the session token, `$data['sessionToken']`.

## How Can It Be Tested?
Use the VSCode PHP debugger to pause code execution before the code acquires the lock.
Open up devtools in the browser, go to the network tab, check "preserve log".
Complete the join flow with the devtools open. When redirected back to the
join form after GoCardless, a request to `join` should appear in the network log.
The execution of this should be paused in VSCode.

Right click the `join` request in the devtools, and click "copy as cURL". Paste and run
this command in a terminal. This should cause another thread to pause before acquiring
the lock.

Resume both threads and check using the join block log that they run sequentially.

## How Will This Be Deployed?
Ask for SSH to be opened, then git pull in the wp-content/ck-join-flow directory. Then, `composer install` in the `packages/join-block` directory.

## Screenshots (if appropriate):

## Types of changes
